### PR TITLE
feat: update otel versions for prometheus to 0.27

### PR DESCRIPTION
**Pull request by kudlatyamroth**
## Changes

- Update `opentelemetry` dependency version to 0.27
- Update `opentelemetry_sdk` dependency version to 0.27
- Update `opentelemetry-semantic-conventions` dependency version to 0.27

## Merge requirement checklist

* [x] [CONTRIBUTING](https://github.com/open-telemetry/opentelemetry-rust/blob/main/CONTRIBUTING.md) guidelines followed
* [x] Unit tests added/updated (if applicable)
* [x] Appropriate `CHANGELOG.md` files updated for non-trivial, user-facing changes
* [ ] Changes in public API reviewed (if applicable)

### CODE REVIEW

**Review by wojciechbator — APPROVED** [review]

File: opentelemetry-prometheus/Cargo.toml [hunk 1]
```diff
@@ -21,17 +21,20 @@ rustdoc-args = ["--cfg", "docsrs"]
 
 [dependencies]
 once_cell = { workspace = true }
-opentelemetry = { version = "0.26", default-features = false, features = ["metrics"] }
-opentelemetry_sdk = { version = "0.26", default-features = false, features = ["metrics"] }
+opentelemetry = { version = "0.27", path = "../opentelemetry", default-features = false, features = ["metrics"] }
```
Comment (cijothomas) [inline opentelemetry-prometheus/Cargo.toml:24]: please remove the path="..". It is intentional until we hit 1.0

Comment (kudlatyamroth) [reply opentelemetry-prometheus/Cargo.toml]: ok, removed :)

File: opentelemetry-prometheus/src/lib.rs [hunk 2]
```diff
@@ -281,7 +273,10 @@ impl prometheus::core::Collector for Collector {
         let mut inner = match self.inner.lock() {
             Ok(guard) => guard,
             Err(err) => {
-                global::handle_error(err);
+                otel_error!(
```
Comment (cijothomas) [inline opentelemetry-prometheus/src/lib.rs:276]: not sure of the impact of this error? is that going to fail entire metrics scraping? if yes, please use a more descriptive message. If not, use a lower severity. Either way, lets try to make sure the internal logs are descriptive enough.

File: opentelemetry-prometheus/src/lib.rs [hunk 3]
```diff
@@ -291,7 +286,10 @@ impl prometheus::core::Collector for Collector {
             scope_metrics: vec![],
         };
         if let Err(err) = self.reader.collect(&mut metrics) {
-            global::handle_error(err);
+            otel_error!(
```
Comment (cijothomas) [inline opentelemetry-prometheus/src/lib.rs:289]: same comment as above.

**Review by cijothomas — CHANGES_REQUESTED** [review]
Changes look good. I am requesting changes to fix the internal logs. These are now emitted via `tracing` and most users likely have a tracing subscriber already enabled.. So we need to be very careful about emitting logs, especially the ones with high severity.
Also, try and make the descriptions more self-explanatory.(Totally understand these are existing issues, but just want to make it easier for end users to navigate the internal logs)

File: opentelemetry-prometheus/src/lib.rs [hunk 4]
```diff
@@ -152,30 +148,26 @@ pub struct PrometheusExporter {
...
         self.reader.force_flush()
     }
 
-    fn shutdown(&self) -> Result<()> {
+    fn shutdown(&self) -> MetricResult<()> {
         self.reader.shutdown()
     }
+
+    fn temporality(&self, kind: InstrumentKind) -> Temporality {
+        self.reader.temporality(kind)
```
[the remark this answers is not in the document]
Comment (kudlatyamroth) [reply opentelemetry-prometheus/src/lib.rs:169]: so, should i hard code it like this:
```
    fn temporality(&self, _kind: InstrumentKind) -> Temporality {
        Temporality::Cumulative
    }
```
?

[the remark this answers is not in the document]
Comment (cijothomas) [reply opentelemetry-prometheus/src/lib.rs:169]: yes, just to make it very obvious that prometheus needs cumulative.

File: opentelemetry-prometheus/src/lib.rs [hunk 5]
```diff
@@ -152,30 +148,26 @@ pub struct PrometheusExporter {
     reader: Arc<ManualReader>,
 }
 
-impl TemporalitySelector for PrometheusExporter {
-    /// Note: Prometheus only supports cumulative temporality so this will always be
-    /// [Temporality::Cumulative].
```
Comment (hdost) [inline opentelemetry-prometheus/src/lib.rs:157]: We should keep this comment on the temporality function

**Review by hdost — CHANGES_REQUESTED** [review]

File: opentelemetry-prometheus/src/lib.rs (hunk 2 above)
Comment (kudlatyamroth) [reply opentelemetry-prometheus/src/lib.rs:278]: im not sure of the impact in these places, but i tried to update description and change two of them to warns :)

Comment (cijothomas) [reply opentelemetry-prometheus/src/lib.rs:278]: Looks like you haven't pushed the commit? Not seeing the change!

Comment (kudlatyamroth) [reply opentelemetry-prometheus/src/lib.rs:278]: i've changed otel error to otel warning in validate metrics, here i think should be error as it results in empty vec.
here i changed only name, end error as a message

Comment (cijothomas) [reply opentelemetry-prometheus/src/lib.rs:278]: > OpenTelemetryPrometheus.AcquireLockError

This name is not end user friendly. Could you use something like MetricScrapeFailed or something similar, avoiding any internal details like locks.

Comment (cijothomas) [reply opentelemetry-prometheus/src/lib.rs:278]: also, there is no need to prefix OpenTelemetryPrometheus - it'll be automatically added as target!

Comment (kudlatyamroth) [reply opentelemetry-prometheus/src/lib.rs:278]: k, changed :)

**Review by cijothomas — APPROVED** [review]

**Review by ttys3 — CHANGES_REQUESTED** [review]

File: opentelemetry-prometheus/Cargo.toml [hunk 6]
```diff
@@ -21,17 +21,20 @@ rustdoc-args = ["--cfg", "docsrs"]
 
 [dependencies]
 once_cell = { workspace = true }
-opentelemetry = { version = "0.26", default-features = false, features = ["metrics"] }
-opentelemetry_sdk = { version = "0.26", default-features = false, features = ["metrics"] }
+opentelemetry = { version = "0.27", default-features = false, features = ["metrics"] }
+opentelemetry_sdk = { version = "0.27", default-features = false, features = ["metrics", "spec_unstable_metrics_views"] }
```
[the remark this answers is not in the document]
Comment (kudlatyamroth) [reply opentelemetry-prometheus/Cargo.toml]: done :)
thanks :)

**Comment by rubenfiszel** [comment]
Hi, is there a published  opentelemetry-prometheus crate that one can depend on with this change?

**Comment by kudlatyamroth** [comment]
i was hoping that merge will be followed by release... but it looks like we will need to wait

**Comment by cijothomas** [comment]
the crate was released last week, sorry missed to update here!

**Comment by cijothomas** [comment]
https://crates.io/crates/opentelemetry-prometheus/0.27.0

**Pull request merged**